src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,174 @@
+        } else {
+            tail.next = newNode;
+            newNode.previous = tail;
+            tail = newNode;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexIsValid(index, size, "Trying to add at nonexistent or negative index");
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            addAtBeginning(value);
+            return;
+        }
+        addInMiddle(value, index);
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        list.forEach(this::add);
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexIsValid(index, size - 1, "Trying to get the value at non "
+                + "existing or negative index");
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexIsValid(index, size - 1, "Trying to set at nonexistent or negative index");
+        Node<T> nodeAtIndex = findNodeByIndex(index);
+        T oldValue = nodeAtIndex.value;
+        nodeAtIndex.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexIsValid(index, size - 1, "Trying to remove at nonexistent or negative index");
+        if (index == 0) {
+            return removeFirstElement();
+        }
+        if (index == size - 1) {
+            return removeLastElement();
+        }
+        return removeFromMiddle(findNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> temp = head;
+        if (head.value == object || head.value.equals(object)) {
+            removeFirstElement();
+            return true;
+        }
+
+        while (temp != null) {
+            if (temp.value == null || temp.value.equals(object)) {
+                removeFromMiddle(temp);
+                return true;
+            }
+            temp = temp.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> previous;
+        private Node<T> next;
+
+        public Node(Node<T> previous, T value, Node<T> next) {
+            this.previous = previous;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private void checkIndexIsValid(int index, int size, String message) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(message);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        int count = 0;
+        Node<T> searchResult = head;
```

Comment:
why are we always starting from head? let's start from tail if our index is closer to tail